### ISSUE TITLE
Lint for functions that don't return but aren't written as "-> !"

### ISSUE THREAD

**Issue by jruderman** [rust-lang/rust#12836]
I think this should warn: f never returns, but it claims to return int.

    fn f() -> int {
      fail!("f")
    }

    fn main() {
      f();
    }

**Comment by nham** [comment]
Triage: still an issue (`rustc 1.1.0-nightly (4b88e8f63 2015-05-11) (built 2015-05-12)`)

FYI the updated code is:

```rust
fn f() -> isize {
  panic!("f")
}

fn main() {
  f();
}
```

**Comment by Toby-S** [comment]
@Manishearth does clippy offer this?

**Comment by Manishearth** [comment]
@Toby-S No, but file a bug. Given that there are some common patterns that encourage this, I'd prefer it to be a pedantic lint (off by default)

**Comment by steveklabnik** [comment]
Since new lints have a big impact on users of rustc, the policy is that they should go through the RFC process like other user-facing changes. As such, I'm going to give this one a close, but if anyone comes across this ticket and wants this lint, consider adding it to clippy and/or writing up an RFC. Thanks!
